addons/web/static/src/core/file_upload/file_upload_service.js

```diff
@@ -0,0 +1,106 @@
+         * @returns {Number}                        upload.loaded
+         * @returns {Number}                        upload.total
+         * @returns {String}                        upload.title
+         * @returns {String||undefined}             upload.type
+         */
+        const upload = async (route, files, params = {}) => {
+            const xhr = this.createXhr();
+            const bus = params.bus || env.bus;
+            xhr.open("POST", route);
+            const formData = new FormData();
+            formData.append("csrf_token", odoo.csrf_token);
+            for (const file of files) {
+                formData.append("ufile", file);
+            }
+            if (params.buildFormData) {
+                params.buildFormData(formData);
+            }
+            const upload = reactive({
+                id: _nextId++,
+                xhr,
+                data: formData,
+                progress: 0,
+                loaded: 0,
+                total: 0,
+                state: "pending",
+                title: files.length === 1 ? files[0].name : sprintf(env._t("%s Files"), files.length),
+                type: files.length === 1 ? files[0].type : undefined,
+            });
+            _uploads[upload.id] = upload;
+            // Progress listener
+            xhr.upload.addEventListener("progress", async (ev) => {
+                upload.progress = ev.loaded / ev.total;
+                upload.loaded = ev.loaded;
+                upload.total = ev.total;
+                upload.state = "loading";
+            });
+            // Load listener
+            xhr.onload = async () => {
+                delete _uploads[upload.id];
+                upload.state = "loaded";
+                bus.trigger("FILE_UPLOAD_LOADED", { upload });
+            };
+            // Error listener
+            xhr.onerror = async () => {
+                delete _uploads[upload.id];
+                upload.state = "error";
+                // Disable this option if you need more explicit error handling.
+                if (params.displayErrorNotification !== undefined && params.displayErrorNotification) {
+                    notificationService.add(env._t("An error occured while uploading."), {
+                        title: env._t("Error"),
+                        sticky: true,
+                    });
+                }
+                bus.trigger("FILE_UPLOAD_ERROR", { upload });
+            };
+            // Abort listener, considered as error
+            xhr.onabort = async () => {
+                delete _uploads[upload.id];
+                upload.state = "abort";
+                bus.trigger("FILE_UPLOAD_ERROR", { upload });
+            };
+            xhr.send(formData);
+            bus.trigger("FILE_UPLOAD_ADDED", { upload });
+            return upload;
+        };
+
+        const getUpload = (id) => _uploads[id];
+        const getUploads = () => _uploads;
```

Comment:
these functions are redundant with one another, the second function also exposes the _uploads outright, at which point a getter is pretty useless, just put the uploads object directly in the returned object.